### ISSUE TITLE
Bokeh serve is not recognised as an internal or external command

### ISSUE THREAD

**Issue by gccrowther** [bokeh/bokeh#5354]
I'm trying to run a bokeh server application using the bokeh serve methodology, after updating to bokeh 12.3, I am now getting the message:

`bokeh` is not recognized as an internal or external command, operable program or batch file.

```
'(server_env) C:\Users\xxxxxxxx\Documents\Python\Projects\bokeh_server>conda list
packages in environment at C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4\envs\server_env:

bokeh                     0.12.3                   py35_0
chest                     0.2.3                    py35_0
cloudpickle               0.2.1                    py35_0
dask                      0.11.1                   py35_0
heapdict                  1.0.0                    py35_1
jinja2                    2.8                      py35_1
locket                    0.2.0                    py35_1
markupsafe                0.23                     py35_2
mkl                       11.3.3                        1
numpy                     1.11.2                   py35_0
pandas                    0.19.0              np111py35_0
partd                     0.3.6                    py35_0
pip                       8.1.2                    py35_0
python                    3.5.2                         0
python-dateutil           2.5.3                    py35_0
pytz                      2016.7                   py35_0
pyyaml                    3.12                     py35_0
requests                  2.11.1                   py35_0
scipy                     0.18.1              np111py35_0
setuptools                27.2.0                   py35_1
six                       1.10.0                   py35_0
toolz                     0.8.0                    py35_0
tornado                   4.4.2                    py35_0
vs2015_runtime            14.0.25123                    0
wheel                     0.29.0                   py35_0

(server_env) C:\Users\xxxxxxxx\Documents\Python\Projects\bokeh_server>bokeh serve --show fbe_image_test.py
'bokeh' is not recognized as an internal or external command,
operable program or batch file.

(server_env) C:\Users\xxxxxxxx\Documents\Python\Projects\bokeh_server>python
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  5 2016, 11:41:13) [MSC v.1900 64 bit (AMD64)] on win32
```

I have uninstalled and reinstalled bokeh from both my environment and top level install using conda, has anyone come across this before?

Thanks,
George

**Comment by bryevdv** [comment]
@gccrowther can you confirm that `bokeh` is not in the `Scripts` dir in the conda env dir?

**Comment by gccrowther** [comment]
Looks like it, this is what I appear to have in there.
'['2to3-script.py',
 '2to3.exe',
 'activate',
 'activate.bat',
 'bokeh',
 'bokeh-server',
 'conda',
 'conda.bat',
 'deactivate',
 'deactivate.bat',
 'easy_install-script.py',
 'easy_install.exe',
 'f2py.py',
 'idle-script.py',
 'idle.exe',
 'pip-script.py',
 'pip.exe',
 'pydoc-script.py',
 'pydoc.exe',
 'wheel-script.py',
 'wheel.exe']`

**Comment by bryevdv** [comment]
@gccrowther I'm not sure, this seems like some kind of windows path issue. Was this window already open when bokeh was installed? Does opening a completely new cmd window or powershell pick it up?

**Comment by mickeydonkey** [comment]
Same problem here when upgrade to 0.12.3 via conda, downgrade to 0.12.2 works fine.

**Comment by mickeydonkey** [comment]
@bryevdv In /Scripts Some bokeh-related files are removed when upgrade to 0.12.3.

**Comment by gccrowther** [comment]
@bryevdv @mickeydonkey Also found that regressing to 0.12.2 works. I notice that in 0.12.3 the 'bokeh' file replaces 'bokeh.exe' and 'bokeh-server-script.py'. 
Only difference I can see in 'bokeh' from the 'bokeh-server-script.py' is the shebang (I think) line to the environment python.exe (which does exist).

**Comment by bryevdv** [comment]
I just confirmed there were no substantive changes to `setup.py` between `0.12.2` and `0.12.3` (just two print statements removed in [this commit](https://github.com/bokeh/bokeh/commit/6aa55d9650f641b0f1d62f9b4499cf6d284f3f48#diff-2eeaed663bd0d25b7e608891384b7298)) Maybe there has been an upstream change.

@gccrowther can you try installing bokeh from the `bokeh` channel on anaconda.org? i.e. `conda install -c bokeh bokeh=0.12.3` instead? What happens if you pip install Bokeh instead into a clean env?

ping @ilanschnell @kalefranz any ideas about this?


@gccrowther @mickeydonkey in the mean time you can use `python -m bokeh`:

```
[bryan:/private/tmp]$ python -m bokeh serve
2016-10-17 07:59:09,336 Starting Bokeh server version 0.12.4dev1-2-gbef6b4a
2016-10-17 07:59:09,346 Starting Bokeh server on port 5006 with applications at paths ['/']
2016-10-17 07:59:09,346 Starting Bokeh server with process id: 28515
```

**Comment by gccrowther** [comment]
@bryevdv 
Pip install into a clean environment does work, I now have a `bokeh.exe` instead of a `bokeh `file.
Conda install from the bokeh channel also works, generating the same `bokeh.exe.`

Also tried `python -m bokeh`, but this yields:
```
(bokeh_server) C:\Users\xxxxxxxx\Documents\Python\Projects>python -m bokeh serve
C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4\envs\bokeh_server\python.exe: No module named bokeh.__main__; 'bokeh' is a package and cannot be directly executed

(bokeh_server) C:\Users\xxxxxxxx\Documents\Python\Projects>python -m bokeh
C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4\envs\bokeh_server\python.exe: No module named bokeh.__main__; 'bokeh' is a package and cannot be directly executed

(bokeh_server) C:\Users\xxxxxxxx\Documents\Python\Projects>python -m bokeh serve fbe_image.py
C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4\envs\bokeh_server\python.exe: No module named bokeh.__main__; 'bokeh' is a package and cannot be directly executed
```
Perfectly viable that I am just using the latter method incorrectly.

**Comment by bryevdv** [comment]
@gccrowther thanks very much for the updated info, that narrows things quite a bit

**Comment by bryevdv** [comment]
@gccrowther @mickeydonkey this should be resolve on   a new conda install or update of `0.12.3` can you confirm?

**Comment by gccrowther** [comment]
@bryevdv 
I'm getting something not dissimilar to the error detailed above. Created a fresh environment and ran `conda install bokeh`. Getting the following:
```
C:\Users\xxxxxxxx\Documents\Python\Projects>conda install bokeh
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bokeh-0.12.3               |           py35_1         3.6 MB

The following packages will be UPDATED:

    bokeh: 0.12.3-py35_0 --> 0.12.3-py35_1

Proceed ([y]/n)? y

Fetching packages ...
bokeh-0.12.3-p 100% |###############################| Time: 0:00:02   1.39 MB/s
Extracting packages ...
[      COMPLETE      ]|##################################################| 100%
Unlinking packages ...
[      COMPLETE      ]|##################################################| 100%
Linking packages ...
[      COMPLETE      ]|##################################################| 100%

C:\Users\xxxxxxxx\Documents\Python\Projects>bokeh serve
Traceback (most recent call last):
  File "C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4\Scripts\bokeh-script.py", line 3, in <module>
    import bokeh.__main__
  File "C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4\Scripts\bokeh.py", line 5, in <module>
    from bokeh.command.bootstrap import main
ImportError: No module named 'bokeh.command'; 'bokeh' is not a package
```

**Comment by bryevdv** [comment]
ping @ilanschnell

**Comment by msarahan** [comment]
@gccrowther I think this is something fishy on your system.  I have made sure that this works in a clean environment:

Sorry @Bryan, this is something fishy on the user's system.

```
C:\Users\builder>conda create -n bokeh_test bokeh python=3.5
Fetching package metadata ...........
Solving package specifications: ..........

Package plan for installation in environment C:\Users\builder\Miniconda2_6\envs
bokeh_test:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    mkl-11.3.3                 |                1       110.0 MB
    vs2015_runtime-14.00.23026.0|                0         2.2 MB  msarahan
    python-3.5.2               |                0        30.3 MB
    markupsafe-0.23            |           py35_2          28 KB
    numpy-1.11.2               |           py35_0         3.3 MB
    pyyaml-3.12                |           py35_0         118 KB
    requests-2.11.1            |           py35_0         665 KB
    setuptools-27.2.0          |           py35_1         761 KB
    six-1.10.0                 |           py35_0           8 KB
    tornado-4.4.2              |           py35_0         593 KB
    wheel-0.29.0               |           py35_0         123 KB
    jinja2-2.8                 |           py35_1         300 KB
    pip-8.1.2                  |           py35_0         1.6 MB
    python-dateutil-2.5.3      |           py35_0         237 KB
    bokeh-0.12.3               |           py35_1         3.6 MB
    ------------------------------------------------------------
                                           Total:       153.8 MB

The following NEW packages will be INSTALLED:

    bokeh:           0.12.3-py35_1
    jinja2:          2.8-py35_1
    markupsafe:      0.23-py35_2
    mkl:             11.3.3-1
    numpy:           1.11.2-py35_0
    pip:             8.1.2-py35_0
    python:          3.5.2-0
    python-dateutil: 2.5.3-py35_0
    pyyaml:          3.12-py35_0
    requests:        2.11.1-py35_0
    setuptools:      27.2.0-py35_1
    six:             1.10.0-py35_0
    tornado:         4.4.2-py35_0
    vs2015_runtime:  14.00.23026.0-0 msarahan
    wheel:           0.29.0-py35_0

Proceed ([y]/n)?

Fetching packages ...
mkl-11.3.3-1.t 100% |###############################| Time: 0:00:16   7.12 MB/s
vs2015_runtime 100% |###############################| Time: 0:00:01   1.28 MB/s
python-3.5.2-0 100% |###############################| Time: 0:00:06   4.84 MB/s
markupsafe-0.2 100% |###############################| Time: 0:00:00 454.68 kB/s
numpy-1.11.2-p 100% |###############################| Time: 0:00:02   1.54 MB/s
pyyaml-3.12-py 100% |###############################| Time: 0:00:00 426.98 kB/s
requests-2.11. 100% |###############################| Time: 0:00:01 680.72 kB/s
setuptools-27. 100% |###############################| Time: 0:00:01 722.76 kB/s
six-1.10.0-py3 100% |###############################| Time: 0:00:00   0.00  B/s
tornado-4.4.2- 100% |###############################| Time: 0:00:00 626.19 kB/s
wheel-0.29.0-p 100% |###############################| Time: 0:00:00 350.33 kB/s
jinja2-2.8-py3 100% |###############################| Time: 0:00:00 531.16 kB/s
pip-8.1.2-py35 100% |###############################| Time: 0:00:01   1.03 MB/s
python-dateuti 100% |###############################| Time: 0:00:00 471.71 kB/s
bokeh-0.12.3-p 100% |###############################| Time: 0:00:02   1.58 MB/s
Extracting packages ...
[      COMPLETE      ]|##################################################| 100%
Linking packages ...
[      COMPLETE      ]|##################################################| 100%
#
# To activate this environment, use:
# > activate bokeh_test
#
# To deactivate this environment, use:
# > deactivate bokeh_test
#
# * for power-users using bash, you must source
#


C:\Users\builder>activate bokeh_test

(bokeh_test) C:\Users\builder>bokeh serve
2016-10-17 18:34:34,181 Starting Bokeh server version 0.12.3
2016-10-17 18:34:34,196 Starting Bokeh server on port 5006 with applications at
paths ['/']
2016-10-17 18:34:34,196 Starting Bokeh server with process id: 2608
```

You should check that no other version of bokeh is getting picked up.  Development versions installed with ```pip install -e``` or ```python setup.py develop``` or ```conda develop``` will take priority over conda installed packages, and it seems like you might be seeing this.

Also, pay attention to what folder you're in.  If there's a bokeh folder in the folder you're running in, python might get confused - especially if there's a ```__init__.py``` file in that folder.

**Comment by gccrowther** [comment]
Apologies gents, that last error is a consequence of me not actually activating my environment. I can confirm that it does now work.
```
(server_test_03) C:\Users\xxxxxxxx\Documents\Python\Projects>conda install bokeh
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment C:\Users\xxxxxxxx\AppData\Local\Continuum\Anaconda3_4\envs\server_test_03:

The following NEW packages will be INSTALLED:

    bokeh:           0.12.3-py35_1
    jinja2:          2.8-py35_1
    markupsafe:      0.23-py35_2
    python-dateutil: 2.5.3-py35_0
    pyyaml:          3.12-py35_0
    requests:        2.11.1-py35_0
    six:             1.10.0-py35_0
    tornado:         4.4.2-py35_0

Proceed ([y]/n)? y

Linking packages ...
[      COMPLETE      ]|##################################################| 100%

(server_test_03) C:\Users\xxxxxxxx\Documents\Python\Projects>bokeh serve
2016-10-18 09:16:57,787 Starting Bokeh server version 0.12.3
2016-10-18 09:16:57,811 Starting Bokeh server on port 5006 with applications at paths ['/']
2016-10-18 09:16:57,812 Starting Bokeh server with process id: 8180
```
Sorry for the confusion, thanks for the assistance.

**Comment by bryevdv** [comment]
great, glad everything is working now

**Comment by gccrowther** [comment]
@bryevdv Thanks for all the help!

**Comment by OttoAbil** [comment]
Hi, I have the same issue. 

I have tried to tun the installation of the development "conda install -c bokeh/label/dev bokeh" thought the command prompt on windows 10. But the the message that it cant install the packages
_conda install -c bokeh/label/dev bokeh
Fetching package metadata ...............
Solving package specifications: .

Package plan for installation in environment C:\Program Files\Anaconda2:

The following packages will be UPDATED:

    bokeh:     0.12.13-py_0          bokeh --> 0.12.14dev3-py_0     bokeh/label/dev
    conda:     4.3.30-py27he817781_0       --> 4.4.7-py27_0
    conda-env: 2.6.0-0                     --> 2.6.0-h36134e3_1
    pycosat:   0.6.2-py27_0                --> 0.6.3-py27hcd410c5_0

Proceed ([y]/n)? y

ERROR conda.core.link:_execute_actions(337): An error occurred while installing package 'bokeh/label/dev::bokeh-0.12.14dev3-py_0'.
CondaError: Cannot link a source that does not exist. C:\Program Files\Anaconda2\Scripts\conda.exe
Running `conda clean --packages` may resolve your problem.
Attempting to roll back.


CondaError: Cannot link a source that does not exist. C:\Program Files\Anaconda2\Scripts\conda.exe
Running `conda clean --packages` may resolve your problem._

**Comment by OttoAbil** [comment]
My solution was running the development version after installing the most recent version of Anaconda
the development is called using this line: 
conda install -c bokeh/channel/dev bokeh

**Comment by bryevdv** [comment]
This seems like it was a conda issue. The recent 4.4 version of conda was a major update. 

Thanks,

Bryan

**Comment by kalefranz** [comment]
The error

    CondaError: Cannot link a source that does not exist. C:\Program Files\Anaconda2\Scripts\conda.exe

related to noarch packages being installed at the same time as a conda update.  It's been fixed in conda 4.3.32 and the fix will also be included in conda 4.4.8.
